src/realm/object-store/c_api/app.cpp

```diff
@@ -834,4 +836,286 @@ RLM_API realm_app_t* realm_user_get_app(const realm_user_t* user) noexcept
     return nullptr;
 }
 
+template <typename T>
+inline util::Optional<T> convert_to_optional(T data)
+{
+    return data ? util::Optional<T>(data) : util::Optional<T>();
+}
+
+template <typename T>
+inline util::Optional<T> convert_to_optional_bson(const realm_string_t* doc)
```

Comment:
I could have coded this in only one method returning always `util::Optional<T>` but I don't like the fact to have to call `value()` on all the other cases where I need to extract the bson object. 
Alternatively, I could have used SFINAE, but I think it overcomplicates the code.